cpp/input/input_manager.cpp

```diff
@@ -0,0 +1,255 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include <algorithm>
+#include <array>
+
+#include "../log/log.h"
+#include "../engine.h"
+#include "input_manager.h"
+
+namespace openage {
+namespace input {
+
+InputManager::InputManager()
+	: // TODO get this from a file instead of hardcoding it here
+	keys {{ sdl_key(SDLK_RETURN), action_t::START_GAME },
+	      { sdl_key(SDLK_ESCAPE), action_t::STOP_GAME },
+	      { sdl_key(SDLK_F1), action_t::TOGGLE_HUD },
+	      { sdl_key(SDLK_F2), action_t::SCREENSHOT },
+	      { sdl_key(SDLK_F3), action_t::TOGGLE_DEBUG_OVERLAY },
+	      { sdl_key(SDLK_F4), action_t::TOGGLE_DEBUG_GRID },
+	      { sdl_key(SDLK_F5), action_t::QUICK_SAVE },
+	      { sdl_key(SDLK_F9), action_t::QUICK_LOAD },
+	      { sdl_key(SDLK_SPACE), action_t::TOGGLE_BLENDING },
+	      { sdl_key(SDLK_F12), action_t::TOGGLE_PROFILER },
+	      { sdl_key(SDLK_m), action_t::TOGGLE_CONSTRUCT_MODE },
+	      { sdl_key(SDLK_p), action_t::TOGGLE_UNIT_DEBUG },
+	      { sdl_key(SDLK_t), action_t::TRAIN_OBJECT },
+	      { sdl_key(SDLK_y), action_t::ENABLE_BUILDING_PLACEMENT },
+	      { sdl_key(SDLK_z, KMOD_LCTRL), action_t::DISABLE_SET_ABILITY },
+	      { sdl_key(SDLK_x, KMOD_LCTRL), action_t::SET_ABILITY_MOVE },
+	      { sdl_key(SDLK_c, KMOD_LCTRL), action_t::SET_ABILITY_GATHER },
+	      { sdl_key(SDLK_g), action_t::SET_ABILITY_GARRISON },
+	      { sdl_key(SDLK_BACKQUOTE), action_t::TOGGLE_CONSOLE},
+	      { sdl_key(SDLK_v), action_t::SPAWN_VILLAGER},
+	      { sdl_key(SDLK_DELETE), action_t::KILL_UNIT},
+	      { sdl_key(SDLK_q), action_t::BUILDING_1},
+	      { sdl_key(SDLK_w), action_t::BUILDING_2},
+	      { sdl_key(SDLK_e), action_t::BUILDING_3},
+	      { sdl_key(SDLK_r), action_t::BUILDING_4},
+	      { sdl_key(SDLK_q, KMOD_LCTRL), action_t::BUILDING_1},
+	      { sdl_key(SDLK_w, KMOD_LCTRL), action_t::BUILDING_2},
+	      { sdl_key(SDLK_e, KMOD_LCTRL), action_t::BUILDING_3},
+	      { sdl_key(SDLK_r, KMOD_LCTRL), action_t::BUILDING_4},
+	      { sdl_key(SDLK_z), action_t::BUILDING_TOWN_CENTER},
+	      { sdl_key(SDLK_1), action_t::SWITCH_TO_PLAYER_1},
+	      { sdl_key(SDLK_2), action_t::SWITCH_TO_PLAYER_2},
+	      { sdl_key(SDLK_3), action_t::SWITCH_TO_PLAYER_3},
+	      { sdl_key(SDLK_4), action_t::SWITCH_TO_PLAYER_4},
+	      { sdl_key(SDLK_5), action_t::SWITCH_TO_PLAYER_5},
+	      { sdl_key(SDLK_6), action_t::SWITCH_TO_PLAYER_6},
+	      { sdl_key(SDLK_7), action_t::SWITCH_TO_PLAYER_7},
+	      { sdl_key(SDLK_8), action_t::SWITCH_TO_PLAYER_8},
+	      { sdl_key(SDLK_UP), action_t::UP_ARROW},
+	      { sdl_key(SDLK_DOWN), action_t::DOWN_ARROW},
+	      { sdl_key(SDLK_LEFT), action_t::LEFT_ARROW},
+	      { sdl_key(SDLK_RIGHT), action_t::RIGHT_ARROW},
+	      { sdl_mouse(1), action_t::PAINT_TERRAIN},
+	      { sdl_mouse(1), action_t::SELECT},
```

Comment:
this is a multimap of hints, mouse click 1 produces an event_arg struct with hints PAINT_TERRAIN and SELECT, it is up to the context it gets passed to, to decide which one gets used.